crates/chain/src/keychain/txout_index.rs

```diff
@@ -580,28 +601,39 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
         &self,
         keychain: &K,
     ) -> Option<SpkIterator<Descriptor<DescriptorPublicKey>>> {
-        let descriptor = self.keychains_to_descriptors.get(keychain)?.1.clone();
-        Some(SpkIterator::new(descriptor))
+        let desc_id = self.keychains_to_descriptor_ids.get(keychain)?;
+        let desc = self
+            .descriptors
+            .get(desc_id)
+            .cloned()
+            .expect("descriptor id cannot be associated with keychain without descriptor");
+        Some(SpkIterator::new(desc))
     }
 
     /// Get unbounded spk iterators for all keychains.
     pub fn all_unbounded_spk_iters(
         &self,
     ) -> BTreeMap<K, SpkIterator<Descriptor<DescriptorPublicKey>>> {
-        self.keychains_to_descriptors
+        self.keychains_to_descriptor_ids
             .iter()
-            .map(|(k, (_, descriptor))| (k.clone(), SpkIterator::new(descriptor.clone())))
+            .map(|(k, desc_id)| {
+                let desc =
+                    self.descriptors.get(desc_id).cloned().expect(
+                        "descriptor id cannot be associated with keychain without descriptor",
+                    );
+                (k.clone(), SpkIterator::new(desc))
+            })
             .collect()
     }
 
     /// Iterate over revealed spks of keychains in `range`
     pub fn revealed_spks(
         &self,
         range: impl RangeBounds<K>,
-    ) -> impl DoubleEndedIterator<Item = (&K, u32, &Script)> + Clone {
-        self.keychains_to_descriptors
+    ) -> impl DoubleEndedIterator<Item = (&K, u32, ScriptBuf)> + Clone {
```

Comment:
What are our thoughts about making keychains non-reference as well?

Reply:
Since K is already `Clone` this sounds like a good idea, it should simplify lifetime issues for users.